File: Programmers/Level2/43165.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

void search(vector<int>& numbers, int target, int idx, int sum, int& caseCnt) {
    if (idx >= numbers.size() - 1) {
        if ((sum + numbers[idx]) == target || (sum - numbers[idx] == target)) {
            ++caseCnt;
        }
        return;
    }
    
    search(numbers, target, idx + 1, sum + numbers[idx], caseCnt);
    search(numbers, target, idx + 1, sum - numbers[idx], caseCnt);
}

int solution(vector<int> numbers, int target) {
    int caseCnt = 0;
    search(numbers, target, 0, 0, caseCnt);
    return caseCnt;
}

File: Programmers/Level2/43165.js
let answer = 0;

function dfs({ target, currentValue, idx, numbers }) {
  if (idx === numbers.length - 1) {
    if ((currentValue + numbers[idx] === target) || (currentValue - numbers[idx] === target)) {
      ++answer;
    }
    return;
  }

  dfs({
    target,
    currentValue: currentValue + numbers[idx],
    idx: idx + 1,
    numbers
  });
  dfs({
    target,
    currentValue: currentValue - numbers[idx],
    idx: idx + 1,
    numbers
  });
}

function solution(numbers, target) {
  let props = {};

  props.target = target;
  props.currentValue = 0;
  props.idx = 0;
  props.numbers = numbers;
  props.answer = 0;

  dfs(props);

  return answer;
}
